src/rust/engine/fs/src/snapshot.rs

```diff
@@ -0,0 +1,352 @@
+    make_file(&dir.path().join(&roland), STR.as_bytes(), 0o600);
+
+    let path_stats = vec![
+      PathStat::file(
+        roland.clone(),
+        downcast_file_stat(posix_fs.stat(roland).unwrap())
+      ),
+    ];
+    // TODO: Inline when only used once
+    let fingerprint = Fingerprint::from_hex_string(
+      "8b1a7ea04eaa2527b35683edac088bc826117b53b7ec6601740b55e20bce3deb",
+    ).unwrap();
+    assert_eq!(
+      Snapshot::from_path_stats(store, digester, path_stats.clone())
+        .wait()
+        .unwrap(),
+      Snapshot {
+        fingerprint: fingerprint,
+        digest: Some(Digest(fingerprint, 78)),
+        path_stats: path_stats,
+      }
+    );
+  }
+
+  #[test]
+  fn snapshot_recursive_directories_including_empty() {
+    let (store, dir, posix_fs, digester) = setup();
+
+    let cats = PathBuf::from("cats");
+    let roland = cats.join("roland");
+    let dogs = PathBuf::from("dogs");
+    let llamas = PathBuf::from("llamas");
+    std::fs::create_dir_all(&dir.path().join(&cats)).unwrap();
+    std::fs::create_dir_all(&dir.path().join(&dogs)).unwrap();
+    std::fs::create_dir_all(&dir.path().join(&llamas)).unwrap();
+    make_file(&dir.path().join(&roland), STR.as_bytes(), 0o600);
+    let dogs_path_stat = PathStat::dir(
+      dogs.clone(),
+      downcast_dir_stat(posix_fs.stat(dogs).unwrap()),
+    );
+    let cats_roland_path_stat = PathStat::file(
+      roland.clone(),
+      downcast_file_stat(posix_fs.stat(roland).unwrap()),
+    );
+    let llamas_path_stat = PathStat::dir(
+      llamas.clone(),
+      downcast_dir_stat(posix_fs.stat(llamas).unwrap()),
+    );
+
+    let path_stats = vec![
+      dogs_path_stat.clone(),
+      cats_roland_path_stat.clone(),
+      llamas_path_stat.clone(),
+    ];
+    let sorted_path_stats = vec![cats_roland_path_stat, dogs_path_stat, llamas_path_stat];
+    // TODO: Inline when only used once
+    let fingerprint = Fingerprint::from_hex_string(
+      "fbff703bdaac62accf2ea5083bcfed89292073bf710ef9ad14d9298c637e777b",
+    ).unwrap();
+    assert_eq!(
+      Snapshot::from_path_stats(store, digester, path_stats.clone())
+        .wait()
+        .unwrap(),
+      Snapshot {
+        fingerprint: fingerprint,
+        digest: Some(Digest(fingerprint, 232)),
+        path_stats: sorted_path_stats,
+      }
+    );
+  }
+
+  struct FileSaver(Arc<Store>, Arc<PosixFS>);
+
+  impl GetFileDigest<String> for FileSaver {
+    fn digest(&self, file: &File) -> BoxFuture<Digest, String> {
+      let file_copy = file.clone();
+      let store = self.0.clone();
+      self
+        .1
+        .clone()
+        .read_file(&file)
+        .map_err(move |err| {
+          format!("Error reading file {:?}: {}", file_copy, err.description())
+        })
+        .and_then(move |content| store.store_file_bytes(&content.content))
+        .to_boxed()
+    }
+  }
+
+  // Ripped from lib.rs's test module
+  // TODO: Work out how to share things across tests
```

Comment:
Thoughts welcome

Reply:
I think that that would require a brand new crate, imported only inside the test module?